doc/source/reference/simd-optimizations.rst

```diff
@@ -111,85 +80,477 @@ The following tables show the current supported optimizations sorted from the lo
+
+**But how this header looks like?**
+
+Let's see how it looks on X86 since the header is dynamically generated
+according to what kind of compiler and architecture we have, we also
+assume the compiler supports these features and it had been successfully
+configured through ``--cpu-baseline="sse sse2 sse3"`` and
+``--cpu-dispatch="ssse3 sse41"``
+
+.. code:: c
+
+   // The header should be located at numpy/numpy/core/src/common/_cpu_dispatch.h
+   /**NOTE
+    ** C defentions that prefixed with "NPY_HAVE_" are representiong
+    ** the required optimzations.
+    **
+    ** C definitions that prefixed with 'NPY__CPU_TARGET_' are protected and
+    ** shouldn't be used by any NumPy C sources.
+    */
+   /******* baseline features *******/
+   /** SSE **/
+   #define NPY_HAVE_SSE 1
+   #include <xmmintrin.h>
+   /** SSE2 **/
+   #define NPY_HAVE_SSE2 1
+   #include <emmintrin.h>
+   /** SSE3 **/
+   #define NPY_HAVE_SSE3 1
+   #include <pmmintrin.h>
+
+   /******* dispatch-able features *******/
+   #ifdef NPY__CPU_TARGET_SSSE3
+     /** SSSE3 **/
+     #define NPY_HAVE_SSSE3 1
+     #include <tmmintrin.h>
+   #endif
+   #ifdef NPY__CPU_TARGET_SSE41
+     /** SSE41 **/
+     #define NPY_HAVE_SSE41 1
+     #include <smmintrin.h>
+   #endif
+
+| **Baseline features** is our minimal set of required optimizations
+  that been configured via
+| ``--cpu-baseline``, They have no preprocessor guards and they're
+  always on, which means they can be used in any source.
+
+   Wait here!! Does NumPy's infrastructure pass the compiler's flags of
+   baseline features to all sources?
+
+Definitely, yes. But wait, the **dispatch-able sources** are treated
+differently.
+
+   What is the **dispatch-able sources**?
+
+Please just continue reading, you will find your answer in the next
+procedure.
+
+   What if the user specifies certain **baseline features** during the
+   build but the running machine doesn't support these kinds of CPU
+   features and at the same time there's instruction-sets laying down in
+   a C source activated by one of these definitions, or maybe the
+   compiler itself auto-generated/vectorized certain piece of code
+   depending on the provided flags?
+
+| Well, during the loading of the NumPy module, there's a validating
+  process detecting
+| this behavior that raises a Python runtime error to inform the user.
+  Otherwise, the CPU/Kernel will interrupt the execution process by
+  raising an illegal instruction error.
+
+| **Dispatch-able features** is our dispatched set of additional
+  optimizations that had been configured via
+| ``--cpu-dispatch``. They're not activated by default and always guard
+  it by other C definitions prefixed with
+| ``NPY__CPU_TARGET_``. C definitions ``NPY__CPU_TARGET_`` are only
+  enabled within **dispatch-able sources**.
+
+6- Dispatch-able sources and configuration statements
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
@mattip, @rgommers, I wonder if such a thing like that can be understandable?
Any ideas!!